Add ability to obtain result of Authentication process

martindufort opened the pull request:
Closes #11.

martindufort committed:
Add scopes to Login object that will store authorized scopes.
Define new closure callback to get result of authentication operation

martindufort committed:
Add new test to ensure we get a proper authenticationResult

martindufort committed:
Update readme to refer to new AuthenticationResult capability
Update test case for AuthenticationResult

mattmassicotte commented:
This is awesome, thank you! I will look more closely soon.

mattmassicotte reviewed (commented):
This is wonderful! Just had two questions about the new function type.

mattmassicotte commented on Sources/OAuthenticator/Authenticator.swift:
```diff
@@ -20,7 +20,8 @@ public enum AuthenticatorError: Error {
 public final class Authenticator {
 	public typealias UserAuthenticator = (URL, String) async throws -> URL
-
+    public typealias AuthenticationResult = (Login?, AuthenticatorError?) -> Void
```
Two questions for this type.

First, what do you think about making the parameter `Result<Login, AuthenticatorError>`? This would help to ease the optional juggling.

Second, I keep coming back to the name `AuthenticationResult`. What do you think about something like `DidLogin` or `LoginHandler`?

martindufort replied:
1. Not sure why but I never use the `Result` type. However I think this is much better. Will make the change.

2. I feel that there are a lot of `Provider` / `Handler` like types in the package. Do you think `DidLogin` would go against the naming convention? Also `LoginHandler` feels that you *need* to handle the login which is not the case. It's only a reporting mechanism. 
What about `AuthenticationStatusHandler` or something like that?

mattmassicotte replied:
1. Great! It took me a while to warm up to `Result` too. And, I often now prefer throwing, since it fits in much better with concurrency. But in this case, it does seem like a good fit.

2. Yeah, I didn't think too hard, but you are making a good point. That name works for me!

martindufort committed:
Rename authentication result callback to use new name: AuthenticationStatusHandler
Update README.md to showcase new usage
Add new test to validate that error in AuthenticationStatusHandler is properly propagated
